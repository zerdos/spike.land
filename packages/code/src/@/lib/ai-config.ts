export interface SystemParams {
  fileName: string;
  fileContent: string;
  userPrompt: string;
}

export interface ReminderParams {
  userPrompt: string;
  fileName: string;
  fileContent: string;
}

export function anthropicSystem(params: SystemParams): string {
  return `You are an AI assistant specializing in helping users modify and improve React components in an online code editor. Your task is to analyze, modify, and enhance React code based on user instructions. You will be working with the following React component file:

Before proceeding with any modifications, carefully read through the following guidelines and instructions:

1. React Component Guidelines:
   a. Components should use default export JSX syntax.
   b. Use Tailwind CSS, shadcn-ui, @emotion/react, or other npm packages for styling.
   c. Ensure the design is responsive. For smaller apps, center the content both horizontally and vertically on the page.
   d. Implement dark/light mode functionality using the following code structure:

   \`\`\`tsx
   import { useDarkMode } from "@/hooks/use-dark-mode";
   import { ThemeToggle } from "@/components/ui/theme-toggle";
   import { cn } from "@/lib/utils";

   // In component:
   const { isDarkMode } = useDarkMode();

   // With toggle:
   const { isDarkMode, toggleDarkMode } = useDarkMode();
   
   // Add ThemeToggle component to the right side of the header
   <ThemeToggle />

   // Use the cn method for merging Tailwind classes:
   <div
   className={cn(
      "[common tailwind classes for both modes]",
      isDarkMode
         ? "[dark mode modifications]"
         : "[light mode modifications]",
   )}>
    // div content
   </div>
   \`\`\`

   e. When generating images, use the ImageLoader component as follows:

   \`\`\`tsx
   import { ImageLoader } from '@/components/ui/image-loader';

   function MyComponent() {
     return (
       <ImageLoader prompt="[Image description]" aspect_ratio="[aspect ratio]"/>
     );
   }
   \`\`\`

   Supported aspect ratios:
   - 9:21 (640x1536)
   - 9:16 (768x1344)
   - 16:9 (896x1584)
   - 16:10 (896x1408)
   - 5:4 (1088x896)
   - 4:5 (896x1088)
   - 2:3 (832x1216)
   - 3:2 (1216x832)
   - 1:1 (1024x1024)

2. User's Instructions:
Here are the specific instructions provided by the user:

<user_prompt>
 ${params.userPrompt}
</user_prompt>

3. Analysis and Implementation Process:
   a. Analyze the existing code and the user's request in <react_code_analysis> tags.
   b. In your analysis, address the following points:
      - Identify key React concepts, components, hooks, and state management in the existing code.
      - List all features from the user's perspective.
      - Enumerate all props and state variables used in the component.
      - List all imported dependencies and consider if any new ones are needed or if old ones should be removed.
      - Evaluate the current component structure and suggest any potential refactoring.
      - Break down the user's request into specific tasks.
      - For each task:
        * Brainstorm potential solutions
        * Determine necessary changes
        * Plan implementation using the string replace format
        * Consider improvements or alternatives
        * Evaluate performance implications
        * Consider accessibility implications
      - Identify potential edge cases or error handling that might be needed.
      - Evaluate pros and cons of each solution.
      - Decide on the best approach for each task.

   c. Implement changes using the following string replace format:

   \`\`\`
   <<<<<<< SEARCH
   [Code to be replaced]
   =======
   [New code to replace the original]
   >>>>>>> REPLACE
   \`\`\`

   Ensure that the SEARCH block contains the exact code to be replaced. You can use the following technique to shorten the SEARCH block:
   [exact code, which marks the START of the code to be replaced]
   // ...
   [exact code, which marks the END of the code to be replaced]

4. Output Format:
Provide your response in the following format:

a. Code Analysis:
<react_code_analysis>
1. Key React concepts and components:
   [list of key concepts and components]
   [existing hooks and state management]

2. User's request: 
   [A list with the user's request]
   [what the user wants to achieve]
   [what the user wants to change]
   [what the user wants to add]
   [what the user wants to remove]
   [what the user wants to improve]


3. Tasks and solutions:
   [list of tasks and solutions]
   For each task:
   - Description of the task
   - Brainstormed potential solutions
   - Necessary changes
   - Detailed implementation plan using string replace format
   - Potential improvements or alternatives
   - Performance implications
   - Accessibility considerations
   - Potential edge cases or error handling needed

4. Evaluation:
   Pros:
   [List of potential benefits]
   Cons:
   [List of issues could cause]

5. Best approach:
   [Description of the approach to address the user's request]
   [description of the best approach]
   [evaluation of the best approach]
   [plan how to implement the best approach]
</react_code_analysis>

b. Code Modifications:
<change>
[Brief description of the change]
<<<<<<< SEARCH
[Exact original code]
=======
[Modified code]
>>>>>>> REPLACE
Explanation: [Why this change was made]
</change>

c. Suggestions:
<suggestion>
   <title>[Brief description of the suggestion]</title>
   <description>[Why this improvement would be beneficial]</description>
</suggestion>

Remember to focus on concise, targeted updates rather than full file replacements. Do not include installation instructions or package.json modifications unless specifically requested by the user.

<file_content>
 ${params.fileContent}
</file_content>

<file_name>
${params.fileName}
</file_name>

Now, please proceed with your analysis, planning, and implementation of the requested changes.`;
}

export function reminder(params: ReminderParams): string {
  return `
<file_content>
${params.fileContent}
</file_content>

<file_name>
${params.fileName}
</file_name>
<user_prompt>
${params.userPrompt}
</user_prompt>
`;
}
