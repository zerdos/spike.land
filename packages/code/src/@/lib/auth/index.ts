/**
 * Unified Authentication Library
 * Provides consistent authentication across all spike.land packages
 *
 * Security Features:
 * - Multiple auth provider support (Clerk, JWT, OAuth)
 * - OWASP compliant security practices
 * - Rate limiting and CSRF protection
 * - Session management with automatic refresh
 * - Audit logging for security events
 * - MFA support
 * - IP whitelisting/blacklisting
 * - Secure password policies
 */

// Export types
export * from "./types";

// Export auth manager
export { AuthManager, createAuthManager } from "./manager";

// Export React hooks
export {
  useAuth,
  useUser,
  useSession,
  useRequireAuth,
  usePermissions,
  useAuthError,
  useMFA,
  useSessionActivity,
  useAuthProvider,
} from "./hooks";

// Export middleware
export {
  createAuthMiddleware,
  createCSRFMiddleware,
  createSessionMiddleware,
  createSecurityHeadersMiddleware,
  createAPIKeyMiddleware,
  createWebhookMiddleware,
  authMiddleware,
  type AuthMiddlewareOptions,
} from "./middleware";

// Export providers
export { BaseAuthProvider } from "./providers/base";
export { ClerkAuthProvider } from "./providers/clerk";
export { JWTAuthProvider } from "./providers/jwt";

// Export security utilities
export { securityUtils } from "./security";

// Default export for convenience
import { createAuthManager } from "./manager";
import { authMiddleware } from "./middleware";

export default {
  createManager: createAuthManager,
  middleware: authMiddleware,
};