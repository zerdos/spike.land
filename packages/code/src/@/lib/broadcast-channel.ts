const broadcastChannels = new Map<string, BroadcastChannel>();

export const getBroadcastChannel = (codeSpace: string) => {
  if (broadcastChannels.has(codeSpace)) {
    return broadcastChannels.get(codeSpace)!;
  }

  const bc = new BroadcastChannel(`codeSpace-${codeSpace}`);

  broadcastChannels.set(codeSpace, bc);

  return bc;
};
