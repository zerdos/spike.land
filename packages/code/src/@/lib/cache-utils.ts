export const CDN_DOMAIN = "cdn.jsdelivr.net";

export const CACHE_VERSION = "1.0.0";

export class CacheUtils {
  static async cleanOldCaches(currentCacheName: string): Promise<void> {
    const cacheNames = await caches.keys();
    await Promise.all(
      cacheNames
        .filter(name => name !== currentCacheName)
        .map(name => caches.delete(name)),
    );
  }

  static setDifference<T>(setA: Set<T>, setB: Set<T>): Set<T> {
    const difference = new Set<T>();
    for (const elem of setA) {
      if (!setB.has(elem)) {
        difference.add(elem);
      }
    }
    return difference;
  }

  static async getMissingFiles(
    missing: Set<string>,
    cacheNames: string[],
    targetCache: Cache,
  ): Promise<Set<string>> {
    const stillMissing = new Set<string>(missing);

    for (const cacheName of cacheNames) {
      if (stillMissing.size === 0) break;

      const cache = await caches.open(cacheName);
      for (const url of stillMissing) {
        const response = await cache.match(url);
        if (response) {
          await targetCache.put(url, response);
          stillMissing.delete(url);
        }
      }
    }

    return stillMissing;
  }

  static async handleCDNRequest(request: Request): Promise<Response> {
    const cache = await caches.open("cdn-cache");
    const cached = await cache.match(request);

    if (cached) {
      return cached;
    }

    const response = await fetch(request);
    if (response.ok) {
      cache.put(request, response.clone());
    }

    return response;
  }

  static async retry<T>(
    fn: () => Promise<T>,
    retries: number = 3,
  ): Promise<T> {
    let lastError: Error | undefined;

    for (let i = 0; i < retries; i++) {
      try {
        return await fn();
      } catch (error) {
        lastError = error as Error;
        if (i < retries - 1) {
          await new Promise(resolve => setTimeout(resolve, 1000 * (i + 1)));
        }
      }
    }

    throw lastError;
  }
}
