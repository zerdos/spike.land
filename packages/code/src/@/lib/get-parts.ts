import type { LanguageMap, ParsingState } from "@/lib/interfaces";

const programmingLanguages: LanguageMap = {
  javascript: ".js",
  typescript: ".ts",
  python: ".py",
  java: ".java",
  c: ".c",
  cpp: ".cpp",
  "c++": ".cpp",
  "c#": ".cs",
  ruby: ".rb",
  php: ".php",
  swift: ".swift",
  "objective-c": ".m",
  kotlin: ".kt",
  go: ".go",
  perl: ".pl",
  rust: ".rs",
  scala: ".scala",
  haskell: ".hs",
  lua: ".lua",
  shell: ".sh",
  sql: ".sql",
  html: ".html",
  css: ".css",
};

const getLanguage = (lang?: string): string => {
  if (!lang) return "plaintext";
  const lowercaseLang = lang.toLowerCase();

  if (["typescript", "ts"].includes(lowercaseLang)) return "typescript";
  if (["javascript", "js"].includes(lowercaseLang)) return "javascript";
  if (["python", "py"].includes(lowercaseLang)) return "python";

  return programmingLanguages[lowercaseLang] || lowercaseLang;
};

const cleanMessageText = (text: string, isUser: boolean): string => {
  if (isUser) {
    // if there is user prompt, return the content of the user prompt
    // <user_prompt>
    // {{userPrompt}}
    // </user_prompt>
    //

    if (text.includes("<user_prompt>")) {
      const parts = text.split("<user_prompt>");
      const userPrompt = parts[1];
      const userPromptParts = userPrompt?.split("</user_prompt>") || [];
      return userPromptParts[0]!.trim();
    }

    return (text || "")
      .split("The user's first message follows:")
      .pop()
      ?.trim()
      .split("Reminder from the system:")[0] || ""
      .trim();
  }
  return text;
};

interface ChatMessagePart {
  type: "text" | "code";
  content: string;
  language?: string;
}

const parseMessageParts = (
  text: string,
  state: ParsingState,
): { parts: ChatMessagePart[]; state: ParsingState; } => {
  const codeBlockRegex = /```(\w+)?\n([\s\S]*?)(?:```|$)/g;
  const parts: ChatMessagePart[] = [];
  let lastIndex = 0;
  let match;

  while ((match = codeBlockRegex.exec(text)) !== null) {
    if (match.index > lastIndex) {
      const textContent = text.slice(lastIndex, match.index);
      if (textContent.trim().length > 0) {
        parts.push({ type: "text", content: textContent.trim() });
      }
    }

    const language = getLanguage(match[1]!);
    const code = match[2]!;

    parts.push({ type: "code", language, content: code });

    lastIndex = codeBlockRegex.lastIndex;
  }

  if (lastIndex < text.length) {
    const remainingText = text.slice(lastIndex);
    if (state.isInCodeBlock) {
      // Inside an incomplete code block
      state.accumulatedContent += remainingText;
      parts.push({
        type: "code",
        language: getLanguage(state.currentLanguage),
        content: state.accumulatedContent,
      });
    } else if (state.isInDiffBlock) {
      // Inside an incomplete diff block
      state.accumulatedDiffContent += remainingText;
      parts.push({
        type: "code",
        language: "diff",
        content: state.accumulatedDiffContent,
      });
    } else {
      // Regular text
      if (remainingText.trim().length > 0) {
        parts.push({ type: "text", content: remainingText.trim() });
      }
    }
  }

  return { parts, state };
};

const extendTextWithDiffMarkers = (
  text: string,
  state: ParsingState,
): string => {
  let result = "";
  let index = 0;

  while (index < text.length) {
    if (state.isInDiffBlock) {
      const endIndex = text.indexOf(">>>>>>> REPLACE", index);
      if (endIndex === -1) {
        // Diff block not yet complete
        state.accumulatedDiffContent += text.slice(index);
        // Wrap the accumulated diff content in a code block without closing it
        result += "```diff\n" + state.accumulatedDiffContent;
        // Do not close the code block; wait for more data
        break;
      } else {
        // Diff block ends here
        state.accumulatedDiffContent += text.slice(
          index,
          endIndex + ">>>>>>> REPLACE".length,
        );
        result += "```diff\n" + state.accumulatedDiffContent + "\n```";
        state.isInDiffBlock = false;
        state.accumulatedDiffContent = "";
        index = endIndex + ">>>>>>> REPLACE".length;
      }
    } else {
      const startIndex = text.indexOf("<<<<<<< SEARCH", index);
      if (startIndex === -1) {
        // No diff block starts; append the rest
        result += text.slice(index);
        break;
      } else {
        // Append text before the diff block
        result += text.slice(index, startIndex);
        index = startIndex;
        // Start of a new diff block
        const endIndex = text.indexOf(">>>>>>> REPLACE", index);
        if (endIndex === -1) {
          // Diff block not yet complete
          state.isInDiffBlock = true;
          state.accumulatedDiffContent = text.slice(index);
          // Wrap the accumulated diff content in a code block without closing it
          result += "```diff\n" + state.accumulatedDiffContent;
          break;
        } else {
          // Diff block ends here

          const diffContent = text.slice(
            index,
            endIndex + ">>>>>>> REPLACE".length,
          );
          result += "```diff\n" + diffContent + "\n```";
          index = endIndex + ">>>>>>> REPLACE".length;
        }
      }
    }
  }

  return result;
};

export const getPartsStreaming = (
  text: string,
  isUser: boolean,
  state: ParsingState = {
    isInCodeBlock: false,
    accumulatedContent: "",
    isInDiffBlock: false,
    accumulatedDiffContent: "",
  },
): { parts: ChatMessagePart[]; state: ParsingState; } => {
  const extendedText = extendTextWithDiffMarkers(text, state);
  const cleanedText = cleanMessageText(extendedText, isUser);
  const { parts, state: newState } = parseMessageParts(cleanedText, state);
  return { parts, state: newState };
};
