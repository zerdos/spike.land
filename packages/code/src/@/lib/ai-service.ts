import { anthropicSystem, reminder } from "@/lib/ai-config";
import type { Message } from "@/lib/interfaces";
import Anthropic from "@anthropic-ai/sdk";
import OpenAI from "openai";

export interface AIServiceConfig {
  anthropicEndpoint: string;
  openAIEndpoint: string;
  gpt4oEndpoint: string;
  updateThrottleMs: number;
  retryWithClaudeEnabled: boolean;
  setIsStreaming: (isStreaming: boolean) => void;
}

export class AIService {
  private _config: AIServiceConfig;

  constructor(config: AIServiceConfig) {
    this._config = config;
  }

  public async sendToAnthropic(
    messages: Message[],
    onUpdate: (code: string) => void,
  ): Promise<Message> {
    this._config.setIsStreaming(true);

    try {
      const client = new Anthropic({
        dangerouslyAllowBrowser: true,
        apiKey: 'DUMMY_API_KEY',
        baseURL: this._config.anthropicEndpoint,
      });

      const stream = await client.messages.create({
        max_tokens: 8192,
        messages: messages.map((message) => ({
          role: message.role === "system" ? "user" : message.role,
          content: message.content.toString(),
        })),
        model: "claude-3-5-sonnet-latest",
        stream: true,
      });

      let content = "";
      for await (const messageStreamEvent of stream) {
        if (
          "type" in messageStreamEvent &&
          messageStreamEvent.type === "content_block_delta" &&
          messageStreamEvent.delta.type === "text_delta"
        ) {
          const chunkContent = messageStreamEvent.delta.text;
          if (chunkContent) {
            content += chunkContent;
            onUpdate(chunkContent);
          }
       
        }
      }

      return {
        id: Math.random().toString(36).substring(7),
        role: "assistant",
        content,
      };
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error("Unknown error occurred");
    } finally {
      this._config.setIsStreaming(false);
    }
  }

  public async sendToGpt4o(
    messages: Message[],
    onUpdate: (code: string) => void,
  ): Promise<Message> {
    this._config.setIsStreaming(true);

    try {
      const client = new OpenAI({
        baseURL: this._config.gpt4oEndpoint,
        apiKey: 'DUMMY_API_KEY',
      });

      const stream = await client.chat.completions.create({
        model: "gpt-4o-mini",
        stream: true,
        messages: messages.map((message) => ({
          role: message.role === "system" ? "user" : message.role,
          content: message.content.toString(),
        })),
      });

      let content = "";
      for await (const chunk of stream) {
        const chunkContent = chunk.choices[0]?.delta?.content;
        if (chunkContent) {
          content += chunkContent;
          onUpdate(chunkContent);
        }
      }

      return {
        id: Math.random().toString(36).substring(7),
        role: "assistant",
        content,
      };
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error("Unknown error occurred");
    } finally {
      this._config.setIsStreaming(false);
    }
  }

  public prepareClaudeContent(
    params: { content: string; messages: Message[]; codeNow: string; codeSpace: string; },
  ): string {
    const { content, messages, codeNow, codeSpace } = params;

    // Check for error messages that should be passed through directly
    if (content.toLowerCase().startsWith("i'm sorry")) {
      return content;
    }

    const lastMessage = messages[messages.length - 1];

    // If codeNow is different from the last message or there are no messages,
    // use anthropic system content
    // if (!lastMessage || lastMessage.content !== codeNow) {
      const result = anthropicSystem({
        fileName: codeSpace+".tsx",
        fileContent: codeNow,
        userPrompt: content,
      });
      return result;
    // }

    // If code hasn't changed, return reminder content
    // return reminder({
    //   userPrompt: content,
    //   fileName: codeSpace+ ".tsx",
    //   fileContent: codeNow,
    // });
  }
}
