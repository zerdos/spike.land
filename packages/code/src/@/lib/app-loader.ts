import { getCodeSpace } from "@/hooks/use-code-space";
import { getCodeSession } from "@/lib/code-session";
import type { ICode } from "@/lib/interfaces";
import { init } from "@/lib/tw-dev-setup";
import { SessionSynchronizer } from "@/services/SessionSynchronizer";

// Centralized type definitions
export interface AppContext {
  codeSpace: string;
  cSess: ICode;
  AppComponent: React.FC<AppComponentProps>;
}

export interface AppComponentProps {
  codeSpace: string;
  cSess: ICode;
}

/**
 * Initializes the application environment
 * Loads modules that should be available globally
 */
export const initializeAppEnvironment = async (): Promise<void> => {
  try {
    const [
      { enhancedFetch },
      { useArchive, useSpeedy },
    ] = await Promise.all([
      import("@/lib/enhanced-fetch"),
      import("@/lib/use-archive"),
    ]);

    // Make utilities available globally for debugging and development
    Object.assign(globalThis, {
      enhancedFetch,
      useArchive,
      useSpeedy,
    });

    console.log("App environment initialized successfully");
  } catch (error) {
    console.error("Error initializing app environment:", error);
    throw error;
  }
};

/**
 * Initializes a WebSocket connection for the specified code space
 */
export const initializeWebSocket = async (codeSpace: string): Promise<void> => {
  if (!codeSpace) {
    throw new Error("Cannot initialize WebSocket: codeSpace is null or undefined");
  }

  try {
    // Import using a relative path to match the project structure
    const { main } = await import("../../ws");
    await main(codeSpace);
    console.log(`WebSocket initialized for code space: ${codeSpace}`);
  } catch (error) {
    console.error("WebSocket initialization failed:", error);
    throw error;
  }
};

/**
 * Loads the application for a given code space
 * This handles initialization of all required components and sessions
 */
export const loadApp = async (pathname: string): Promise<AppContext | null> => {
  const codeSpace = getCodeSpace(pathname);

  if (!codeSpace) {
    console.error("Cannot load app: Invalid codeSpace from pathname", pathname);
    return null;
  }

  try {
    // Initialize tailwind
    await init();

    // Get code session
    const cSess = await getCodeSession(codeSpace);

    // Load the app component dynamically
    // Import using a relative path to match the project structure
    const { AppToRender } = await import("../../AppToRender");

    // Initialize app environment
    await initializeAppEnvironment();

    return {
      codeSpace,
      cSess,
      AppComponent: AppToRender,
    };
  } catch (error) {
    console.error("Error loading app:", error);
    return null;
  }
};

/**
 * Initializes a session synchronizer for the given code space and session
 */
export const initializeSessionSync = async (
  codeSpace: string,
  cSess: ICode,
): Promise<() => void> => {
  if (!codeSpace || !cSess) {
    console.error("Cannot initialize session sync: missing codeSpace or cSess");
    return () => {};
  }

  try {
    // Make cSess available globally for debugging
    Object.assign(globalThis, { cSess });

    // Initialize session synchronizer
    const sessionSync = new SessionSynchronizer(codeSpace);
    await sessionSync.init(await cSess.getSession());

    // Return the unsubscribe function
    return sessionSync.subscribe((sess) => {
      cSess.setSession(sess);
      // Note: setState callback would be handled by the component
    });
  } catch (error) {
    console.error("Error initializing session sync:", error);
    return () => {};
  }
};

/**
 * Utility to check if a route should initialize WebSocket
 */
export const shouldInitWebSocket = (pathname: string, codeSpace: string | null): boolean => {
  if (!pathname || !codeSpace) return false;
  return pathname === `/live/${codeSpace}/iframe`;
};
