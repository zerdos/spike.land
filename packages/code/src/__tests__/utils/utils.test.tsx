import { getPartsStreaming } from "@/lib/get-parts";
import { describe, expect, test } from "vitest";

describe("getParts", () => {
  test("should handle empty input", () => {
    const input = "";
    const result = getPartsStreaming(input, true);
    expect(result.parts).toMatchInlineSnapshot([]);
  });

  test("should handle text without code blocks", () => {
    const input = "This is a simple text message.";
    const result = getPartsStreaming(input, true);
    expect(result.parts).toMatchInlineSnapshot([
      {
        type: "text",
        content: "This is a simple text message.",
      },
    ]);
  });

  test("should handle text with a single code block", () => {
    const input = "Here is some code:\n```typescript\nconst x = 5;\nconsole.log(x);\n```";
    const result = getPartsStreaming(input, false);
    expect(result.parts).toMatchInlineSnapshot(`
      [
        {
          "content": "Here is some code:",
          "type": "text",
        },
        {
          "content": "const x = 5;
      console.log(x);
      ",
          "language": "typescript",
          "type": "code",
        },
      ]
    `);
  });

  test("should handle text with multiple code blocks", () => {
    const input =
      'First block:\n```javascript\nlet a = 1;\n```\nSecond block:\n```python\nprint("Hello")\n```';
    const result = getPartsStreaming(input, true);
    expect(result).toMatchInlineSnapshot(`
      {
        "parts": [
          {
            "content": "First block:",
            "type": "text",
          },
          {
            "content": "let a = 1;
      ",
            "language": "javascript",
            "type": "code",
          },
          {
            "content": "Second block:",
            "type": "text",
          },
          {
            "content": "print("Hello")
      ",
            "language": "python",
            "type": "code",
          },
        ],
        "state": {
          "accumulatedContent": "",
          "accumulatedDiffContent": "",
          "isInCodeBlock": false,
          "isInDiffBlock": false,
        },
      }
    `);
  });

  test("should handle code blocks with no language specified", () => {
    const input = "Code without language:\n```\nsome code\n```";
    const result = getPartsStreaming(input, false);
    expect(result.parts).toMatchInlineSnapshot(`
      [
        {
          "content": "Code without language:",
          "type": "text",
        },
        {
          "content": "some code
      ",
          "language": "plaintext",
          "type": "code",
        },
      ]
    `);
  });

  test("should handle search and replace markers", () => {
    const input = "Before\n<<<<<<< SEARCH\nold code\n=======\nnew code\n>>>>>>> REPLACE\nAfter";
    const result = getPartsStreaming(input, true);
    expect(result.parts).toMatchSnapshot();
  });

  test("should handle nested code blocks", () => {
    const input =
      'Outer block:\n```javascript\nfunction example() {\n  console.log(`Inner block:\n```json\n{"key": "value"}\n````);\n}\n```';
    const result = getPartsStreaming(input, true);
    expect(result).toMatchSnapshot();
  });

  test("should handle an incomplete code block at the end", () => {
    const input = "Some text\n```javascript\nlet x = 10;\nconsole.log(x);";
    const result = getPartsStreaming(input, false);
    expect(result).toMatchInlineSnapshot(`
      {
        "parts": [
          {
            "content": "Some text",
            "type": "text",
          },
          {
            "content": "let x = 10;
      console.log(x);",
            "language": "javascript",
            "type": "code",
          },
        ],
        "state": {
          "accumulatedContent": "",
          "accumulatedDiffContent": "",
          "isInCodeBlock": false,
          "isInDiffBlock": false,
        },
      }
    `);
  });

  test("should handle multiple search and replace markers", () => {
    const input = `
    first change:
      <<<<<<< SEARCH
      const a = 1;
      =======
      const a = 2;
      >>>>>>> REPLACE
    second change:
      <<<<<<< SEARCH
      let b = 3;
      =======
      let b = 4;
      >>>>>>> REPLACE
    `;

    const result = getPartsStreaming(input, true);
    expect(result).toMatchSnapshot();
  });

  test("should handle incomplete search and replace markers", () => {
    const input = `
    first change:
      <<<<<<< SEARCH
      const a = 1;
      =======
      const a = 2;
      =======
    second change:
      <<<<<<< SEARCH
      let b = 3;
      =======
      let b =
    `;

    const result = getPartsStreaming(input, false);
    expect(result.parts[0]).toMatchSnapshot();
    expect(result.parts[1]).toMatchSnapshot();
  });
});
