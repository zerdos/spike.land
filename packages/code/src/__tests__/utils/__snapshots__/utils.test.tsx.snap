// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getParts > should handle incomplete search and replace markers 1`] = `
{
  "content": "first change:",
  "type": "text",
}
`;

exports[`getParts > should handle incomplete search and replace markers 2`] = `
{
  "content": "<<<<<<< SEARCH
      const a = 1;
      =======
      const a = 2;
      =======
    second change:
      <<<<<<< SEARCH
      let b = 3;
      =======
      let b =
    ",
  "language": "diff",
  "type": "code",
}
`;

exports[`getParts > should handle multiple search and replace markers 1`] = `
{
  "parts": [
    {
      "content": "first change:",
      "type": "text",
    },
    {
      "content": "<<<<<<< SEARCH
      const a = 1;
      =======
      const a = 2;
      >>>>>>> REPLACE
",
      "language": "diff",
      "type": "code",
    },
    {
      "content": "second change:",
      "type": "text",
    },
    {
      "content": "<<<<<<< SEARCH
      let b = 3;
      =======
      let b = 4;
      >>>>>>> REPLACE
",
      "language": "diff",
      "type": "code",
    },
  ],
  "state": {
    "accumulatedContent": "",
    "accumulatedDiffContent": "",
    "isInCodeBlock": false,
    "isInDiffBlock": false,
  },
}
`;

exports[`getParts > should handle nested code blocks 1`] = `
{
  "parts": [
    {
      "content": "Outer block:",
      "type": "text",
    },
    {
      "content": "function example() {
  console.log(\`Inner block:
",
      "language": "javascript",
      "type": "code",
    },
    {
      "content": "json
{"key": "value"}
\`\`\`\`);
}
\`\`\`",
      "type": "text",
    },
  ],
  "state": {
    "accumulatedContent": "",
    "accumulatedDiffContent": "",
    "isInCodeBlock": false,
    "isInDiffBlock": false,
  },
}
`;

exports[`getParts > should handle search and replace markers 1`] = `
[
  {
    "content": "Before",
    "type": "text",
  },
  {
    "content": "<<<<<<< SEARCH
old code
=======
new code
>>>>>>> REPLACE
",
    "language": "diff",
    "type": "code",
  },
  {
    "content": "After",
    "type": "text",
  },
]
`;
